### DISCUSSION TITLE
Unable to load SuperSet UI when I'm using via a domain

### DISCUSSION THREAD

**Discussion by vikashrajgupta** [apache/superset#27924] [Q&A / Help]
### Bug description

I'm getting the below error when using a domain name.

```
Refused to load the script 'https://superset1.greyb.com/cdn-cgi/scripts/7d0fa10a/cloudflare-static/rocket-loader.min.js' because it violates the following Content Security Policy directive: "script-src 'self' 'strict-dynamic' 'nonce-PGDBKAzYO-8kvbduD-2u4qVUJAaWVbte'". Note that 'strict-dynamic' is present, so host-based allowlisting is disabled. Note that 'script-src-elem' was not explicitly set, so 'script-src' is used as a fallback.

GET https://static.cloudflareinsights.com/beacon.min.js/v84a3a4012de94ce1a686ba8c167c359c1696973893317 net::ERR_BLOCKED_BY_CLIENT
```

Firstly I set up this project locally and then on server on both places it works fine but using IP, whenever I'm using a domain name it shows the above error.

The login page is loading perfectly after the username and password UI is not visible., I tried to update config.py and superset_config.py by every possibility but nothing works.

### How to reproduce the bug

1. opening superset1.greyb.com
2. Go to the login page (you will be able to see the error in the console)
3. After entering a username and password it will redirect to interfaces that are not loading

### Screenshots/recordings

![Screenshot from 2024-04-04 18-45-35](https://github.com/apache/superset/assets/40394708/cabbefe0-39b1-4afb-b330-874b20a072ce)

### Superset version

master / latest-dev

### Python version

3.10

### Node version

18 or greater

### Browser

Chrome

### Additional context

I'm facing this issue since last 3-4 days, that would be great if you can help me to resolve this.

### Checklist

- [X] I have searched Superset docs and Slack and didn't find a solution to my problem.
- [X] I have searched the GitHub issue tracker and didn't find a similar bug report.
- [ ] I have checked Superset's logs for errors and if I found a relevant Python stacktrace, I included it here as text in the "additional context" section.

**Comment by rusackas** [comment]
Going to the provided url doesn't show Superset at all:
<img width="855" alt="image" src="https://github.com/apache/superset/assets/812905/3f3b16e7-bbc2-4ddc-b464-4a4e26ab15e0">

That said, whatever `rocket-loader` is, it's being blocked by Superset's CSP. You can add that domain to your `TALISMAN_CONFIG` configuration in `config.py`

**Comment by vikashrajgupta** [comment]
@rusackas - Sorry for the incomplete details that I provided earlier, also I have tried to change the `TALISMAN_CONFIG` in my config.py

superset1.greyb.com is hosted on my internal network and it's not public yet, I have also confirmed and there is no blocking or any kind of issue from our end.

Here is my `TALISMAN_CONFIG` for both production and development mode.

```
TALISMAN_CONFIG = {
    'contentSecurityPolicy': False,
    "content_security_policy": {
        "default-src": ["'self'","https://superset1.greyb.com"],
        "img-src": ["'self'", "data:","https://superset1.greyb.com"],
        "worker-src": ["'self'", "blob:","https://superset1.greyb.com"],
        "connect-src": [
            "'self'",
            "https://api.mapbox.com",
            "https://events.mapbox.com",
            "https://superset1.greyb.com",
        ],
        "object-src": "'none'",
        "style-src": ["'self'", "'unsafe-inline'","'https://superset1.greyb.com'","https://superset1.greyb.com/cdn-cgi/scripts/7d0fa10a/cloudflare-static/rocket-loader.min.js"],
        "script-src": ["'self'", "'unsafe-inline'","'unsafe-eval'","'https://superset1.greyb.com'"],
    },
    "content_security_policy_nonce_in": ["script-src"],
    "force_https":False,
}
```
I have tried many things and option related to this but none of them is working. and whenever I'm accessing with my IP its working fine as shown below.

**Comment by vikashrajgupta** [comment]
@rusackas - I have a request, do not close this until I get a solution from your side, and we're very curious to know the possible solution from your side. 
As I have mentioned the configuration above, I just wanted to share I have tried many permutation combinations to resolve this issue but I failed.
Thanks!!!

**Comment by rusackas** [comment]
It looks like you have some weird double quoting in your config there, e.g. `"'https://superset1.greyb.com'"` should be `"https://superset1.greyb.com"`. From the error you posted, it also seems that it's trying to do a GET to `static.cloudflareinsights.com` so you'll want to punch a hole for that too.

This doesn't seem like a bug with Superset, but rather a config issue... if you want to keep it open, I can move it to a GitHub discussion in the Q&A section.

**Comment by vikashrajgupta** [comment]
@rusackas - Tried with this config as well getting same error.
```
TALISMAN_CONFIG = {
    'contentSecurityPolicy': False,
    "content_security_policy": {
        "default-src": ["'self'","https://superset1.greyb.com"],
        "img-src": ["'self'", "data:","https://superset1.greyb.com"],
        "worker-src": ["'self'", "blob:","https://superset1.greyb.com"],
        "connect-src": [
            "'self'",
            "https://api.mapbox.com",
            "https://events.mapbox.com",
            "https://superset1.greyb.com",
        ],
        "object-src": "'none'",
        "style-src": ["'self'", "'unsafe-inline'","https://superset1.greyb.com"],
        "script-src": ["'self'", "'unsafe-inline'","'unsafe-eval'","https://superset1.greyb.com"],
    },
    "content_security_policy_nonce_in": ["script-src"],
    "force_https":False,
}
```
As I have also mentioned tried a few alternative and a combination as well

**Comment by vikashrajgupta** [comment]
Do you have any suggestions to resolve this issue? because I also tried to do some changes in superset_config.py but didn't work.

**Comment by rusackas** [comment]
Well, there might be a few problems:
* You have both `contentSecurityPolicy` and `content_security_policy` in there. Get rid of `contentSecurityPolicy`
* It still doesn't look like you've allowed `https://static.cloudflareinsights.com`
* Are you running in dev or prod mode? Note that there are two configs... TALISMAN_CONFIG and TALISMAN_DEV_CONFIG, so you should be using the right one.

**Reply by vikashrajgupta** [reply]
@rusackas - I have removed the `contentSecurityPolicy` as per your suggestion

I have also allowed static.cloudflareinsights

I'm running in production mode but for the safe side, I have kept both configs the same.
```
TALISMAN_DEV_CONFIG = {
    "content_security_policy": {
        "default-src": ["'self'", "https://superset1.greyb.com","https://superset1.greyb.com","https://static.cloudflareinsights.com"],
        "img-src": ["'self'", "data:"],
        "worker-src": ["'self'", "blob:","https://superset1.greyb.com","https://static.cloudflareinsights.com"],
        "connect-src": [
            "'self'",
            "https://api.mapbox.com",
            "https://events.mapbox.com",
            "https://superset1.greyb.com",
            "https://static.cloudflareinsights.com",
        ],
        "object-src": "'none'",
        "style-src": ["'self'", "'unsafe-inline'","https://superset1.greyb.com","https://static.cloudflareinsights.com"],
        "script-src": ["'self'", "'unsafe-inline'", "'unsafe-eval'","https://superset1.greyb.com","https://static.cloudflareinsights.com"],
    },
    "content_security_policy_nonce_in": ["script-src"],
    "force_https": False,
}
```
Updated the config file and then stopped docker and restarted as well but didn't work.

**Reply by vikashrajgupta** [reply]
@rusackas - Waiting for your response, that would be great if you can help me in this

**Reply by rusackas** [reply]
Sorry, at this point, I'm not sure. Are ANY of those talisman configs working? Just for testing's sake, you could try disabling talisman entirely via TALISMAN_ENABLED, to see if that makes a difference.

**Reply by vikashrajgupta** [reply]
@rusackas - I have tried enabling and disabling  TALISMAN_ENABLED but didn't work in my case, also the same config is working fine when I'm accessing the same via http://localhost:9888/superset/welcome/  but with the domain name is not working and showing the console error.

Please correct me If I'm doing the correct things by enabling and disabling the `TALISMAN_ENABLED` using true and false value.
 here is the updated config
```
TALISMAN_ENABLED=True

# React requires `eval` to work correctly in dev mode
TALISMAN_DEV_CONFIG = {
    "content_security_policy": {
        "default-src": ["'self'", "https://superset1.greyb.com","https://superset1.greyb.com","https://static.cloudflareinsights.com"],
        "img-src": ["'self'", "data:","https://static.cloudflareinsights.com"],
        "worker-src": ["'self'", "blob:","https://superset1.greyb.com","https://static.cloudflareinsights.com"],
        "connect-src": [
            "'self'",
            "https://api.mapbox.com",
            "https://events.mapbox.com",
            "https://superset1.greyb.com",
            "https://static.cloudflareinsights.com",
        ],
        "object-src": "'none'",
        "style-src": ["'self'", "'unsafe-inline'","https://superset1.greyb.com","https://static.cloudflareinsights.com"],
        "script-src": ["'self'", "'unsafe-inline'", "'unsafe-eval'","https://superset1.greyb.com","https://static.cloudflareinsights.com"],
    },
    "content_security_policy_nonce_in": ["script-src"],
    "force_https": False,
}
```

**Reply by rusackas** [reply]
Yes, `TALISMAN_ENABLED` is a boolean feature flag. It's true by default for good reason. Turning it off should prevent **_any_** CSP checking from taking place. If that's not resolving your problem, then it sounds like your config isn't being picked up at all, for reasons I'm completely unsure of. Alternatively, maybe you could just disable rocket-loader and not even have this problem? Last but not least, you might not want the default admin user to have access on a public instance ;)

**Reply by vikashrajgupta** [reply]
@rusackas - Earlier, it was false, although I have tried it with true and false as well, and it didn't work.
Yes, you're right, The default admin user is maintained by us; it will not be available to the public, but we need this too.
We will provide a public instance to our users with a few restrictions, but the problem is that it's not even opening in the public domain.
`just disable rocket-loader and not even have this problem`->> I'm not sure if disabling rocket-loader will work perfectly in my case.
Should I have to add any configuration to disable this?

**Reply by vikashrajgupta** [reply]
@rusackas - Disabling rocket-loader worked in this case, I am just curious to know why this issue came while shifting to a domain name.
`GET https://static.cloudflareinsights.com/beacon.min.js/v84a3a4012de94ce1a686ba8c167c359c1696973893317 net::ERR_BLOCKED_BY_CLIENT`->> I'm also getting this, wanted to know if this will create an issue anywhere in the superset.

**Reply by rusackas** [reply]
Looks like the same issue... cloudflareinsights.com is probably not enabled in your CSP.

**Reply by vikashrajgupta** [reply]
Thanks @rusackas disabling the rocket loader worked in my case but I just wanted to share there are more issues related to UI in the latest version i.e. 4.0.1 that's why for now I'm using 3.0.1

UI Issue in Iframe - In Iframe superset content is not loading and showing `it's blocked by your origin` but in the older version, it is working fine.


Thanks! Again for the help!!
